### PR TITLE
[WIP] Refactor old interface code to a single library

### PR DESCRIPTION
**Pull request by DrakBoul**
This PR is in regards to #61

In this PR I refactor all our networking and hardware interfaces into a single, well organized library. TCP, UART, I2C now all use the same "Interface" defined in interfaces/lib.rs, icp does not share this trait as it differs quite a bit from the other interfaces we have. With the refactor came changing the paths for modules and crates in each toml and most main's in the repo. Another notable change is that I changed "send" function to "write" in the interface definition as I found this to be the proper nomenclature for what each interface actually does. Please give me feedback on if I should keep this or revert back to "send".

Using interactive rebase I sqaushed commits for deleting old interface libs, adding new interface library, and modifying toml's and main's to have the right path all into a single commit. @rcunrau please let me know if this is proper use of "squashing", or if this commit is so large that it is no longer managable to review.

Note that I still have to adjust some of the README's according to the changes I made.

### CODE REVIEW

**Comment by rcunrau** [comment]
I think we may actually be twins separated at birth or something, since we apparently have the same thoughts at the same time. Did you see my [PR](https://github.com/AlbertaSat/ex3_software/pull/69), which refactors both the common and interface libraries? We do slightly different things, but there's a *lot* of overlap, so we're going to have to work out a plan to combine them.

File: ex3_obc_fsw/handlers/adcs_handler/src/main.rs [hunk 1]
```diff
@@ -246,7 +245,7 @@ impl ADCSHandler {
 
     fn send_cmd(&mut self, command: sim_adcs::ADCSCmdParam, msg: Msg) -> Result<(), Error> {
         let cmd = self.build_cmd(command, msg)?;
-        self.peripheral_interface.as_mut().unwrap().send(&cmd)?;
+        self.peripheral_interface.as_mut().unwrap().write(&cmd)?;
```
Comment (rcunrau) [inline ex3_obc_fsw/handlers/adcs_handler/src/main.rs:248]: To me, read belongs with write, and send belongs with recv :-)

About squashing though, I would say all the path and use changes can be squashed together, but all the send->write changes should be squashed together in a different commit.

Comment (DrakBoul) [reply ex3_obc_fsw/handlers/adcs_handler/src/main.rs:248]: That makes sense, if I didn't squash that commit I could have reverted back to make changes easier. I'll keep that in mind for the future. As for the naming conventions, I am terribly sorry but you have mentioned using write (in place of read) and recv (in place of send) a couple times now and it still does not make sense to me :(. We should have a whiteboard session so we can make sense of this with diagrams, because to me what your suggesting sounds backwards.

**Comment by DrakBoul** [comment]
> I think we may actually be twins separated at birth or something, since we apparently have the same thoughts at the same time. Did you see my [PR](https://github.com/AlbertaSat/ex3_software/pull/69), which refactors both the common and interface libraries? We do slightly different things, but there's a _lot_ of overlap, so we're going to have to work out a plan to combine them.

Haha I did not realize you made that PR, I haven't taken a look yet but I would say that we should keep yours since it does what my PR did and more.

**Comment by rcunrau** [comment]
I merged [PR](https://github.com/AlbertaSat/ex3_software/pull/69) but everything here is worthwhile. I agree tcp_interface shouldn't be in common, and the interface trait is a great idea.

**Comment by DrakBoul** [comment]
> I merged [PR](https://github.com/AlbertaSat/ex3_software/pull/69) but everything here is worthwhile. I agree tcp_interface shouldn't be in common, and the interface trait is a great idea.

Perfect, I realized after the fact that it might be better to keep my stuff in this PR. I'll be busy with family stuff this weekend, so I'll finish up #68 on sunday night.

**Comment by DrakBoul** [comment]
Closing this PR because solving all the merge conflicts would take longer than starting from scratch.
